graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphQLService.java

```diff
@@ -0,0 +1,233 @@
+        if (Strings.isNullOrEmpty(value)) {
+            return ImmutableMap.of();
+        }
+        return parseJsonString(value);
+    }
+
+    private HttpResponse execute(ServiceRequestContext ctx, ExecutionInput input) {
+        if (useBlockingTaskExecutor) {
+            final CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+            ctx.blockingTaskExecutor().execute(() -> {
+                try {
+                    final ExecutionResult executionResult = graphQL.execute(input);
+                    future.complete(toHttpResponse(executionResult));
+                } catch (RuntimeException e) {
+                    final ExecutionResult error = toExecutionResult(e);
+                    future.complete(HttpResponse.of(MediaType.JSON_UTF_8,
+                                                    toJsonString(error.toSpecification())));
+                }
+            });
+            return HttpResponse.from(future);
+        }
+
+        return HttpResponse.from(graphQL.executeAsync(input).handle((executionResult, cause) -> {
+            if (cause != null) {
+                final ExecutionResult error = toExecutionResult(cause);
+                return HttpResponse.of(MediaType.JSON_UTF_8, toJsonString(error.toSpecification()));
+            }
+            return toHttpResponse(executionResult);
+        }));
+    }
+
+    private static HttpResponse toHttpResponse(ExecutionResult executionResult) {
+        // TODO: When WebSocket is implemented, it should be removed.
+        if (executionResult.getData() instanceof Publisher) {
+            final ExecutionResult error =
+                    toExecutionResult(new UnsupportedOperationException("WebSocket is not implemented"));
+            return HttpResponse.of(MediaType.JSON_UTF_8, toJsonString(error.toSpecification()));
+        }
+        return HttpResponse.of(MediaType.JSON_UTF_8, toJsonString(executionResult.toSpecification()));
+    }
+
+    private static ExecutionResult toExecutionResult(Throwable cause) {
+        return new ExecutionResultImpl(GraphqlErrorException.newErrorException()
+                                                            .message(cause.getMessage())
+                                                            .cause(cause)
+                                                            .build());
+    }
+
+    private static Map<String, Object> parseJsonString(String content) {
+        try {
+            return OBJECT_MAPPER.readValue(content, JSON_MAP);
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException("failed to parse a JSON document: " + content, e);
+        }
+    }
+
+    private static String toJsonString(Map<String, Object> result) {
+        try {
+            return OBJECT_MAPPER.writeValueAsString(result);
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException("failed to write a JSON document: " + result, e);
```

Comment:
`IllegalArgumentException` is usually used for `404 Bad Request`. However, this method is used for encoding a response.
Could you return `IllegalStateException` when failing to write `result` as a string?